File: codechef/contests/SEPT16/CHMTDV.py
'''input
6 3
1 1 1 1 1 5
1 1 1 1 1 5
1 1 1 1 1 5
1 1 1 1 1 5
1 1 1 1 1 5
1 1 1 1 1 5
'''

import os, sys, collections, itertools, pprint, fractions, math, random

ai = lambda : map(int, raw_input().split())
ii = lambda : int(raw_input())

n,p = None, None

def fits(mid, Sum, V):
	'''Returns 1 if need to increase sum i.e., lo = mid+1
	Returns 2 if need to decrease the sum i.e., hi = mid'''
	del V[:]
	cursum = 0
	for j in xrange(n):
		if Sum[j] > mid: return 1
		if cursum+Sum[j] > mid:
			V.append(j)
			cursum = Sum[j]
		else:
			cursum += Sum[j]
	if len(V) > p-1: return 1
	if len(V) < p-1: return 3
	return 2

def fits2(mid, FM, V):
	cursum = [0]*p
	del V[:]
	for i in xrange(len(FM)):
		for j in xrange(len(FM[0])):
			if FM[i][j] > mid: return 1
	for i in xrange(n):
		for j in xrange(p):
			if cursum[j]+FM[j][i] > mid:
				V.append(i)

				cursum = [FM[k][i] for k in xrange(p)]
				break
			else:
				cursum[j] += FM[j][i]
	if len(V) < p-1:
		return 3 # hi = mid-1
	if len(V) > p-1:
		return 1
	return 2


n, p = ai()
a = []
for _ in xrange(n):
	a.append(ai())

rsum = [sum(a[i]) for i in xrange(n)]

lo, mid, hi = 0, 0, 10**18
H, V = [], []
while lo <= hi:
	if mid == (lo+hi)/2: break
	mid = (hi+lo)/2
	ans = fits(mid, rsum, H)
	# print "H:",H
	if ans == 2: hi = mid
	elif ans == 3: hi = mid-1
	else: lo = mid+1

if lo > hi:
	lo, hi = hi, lo
if (fits(lo, rsum, H) == 2):
	None
else:
	fits(hi, rsum, H)

FM = [[0 for i in xrange(n)] for j in xrange(p)]

for col in xrange(n):
	for row in xrange(p-1):
		if row == 0:
			for crow in xrange(H[row]):
				FM[row][col] += a[crow][col]
		else:
			for crow in xrange(H[row-1],H[row]):
				FM[row][col] += a[crow][col]
			# if row != 0 and row == p-2:
			# 	for crow in xrange(H[row],n):
			# 		FM[row+1][col] += a[crow][col]
		if row == p-2:
			for crow in xrange(H[row],n):
				FM[row+1][col] += a[crow][col]



lo, mid, hi = 0, 0, 10**18
while lo <= hi:
	if mid == (lo+hi)/2: break
	mid = (lo+hi)/2
	ans = fits2(mid, FM, V)
	if ans == 3:
		hi = mid-1
	elif ans == 2:
		hi = mid
	else:
		lo = mid+1

if lo > hi: lo, hi = hi, lo

if fits2(lo, FM, V):
	None
else:
	fits2(hi, FM, V)

TV = []
maxelem = max(max(FM[i]) for i in xrange(p))
avg = sum(FM[0])/p
for i in xrange(1,p):
	avg = max(sum(FM[i])/p, avg)
fits2(max(avg, maxelem), FM, TV)

for s in H:
	print s,
print

if len(V) == len(TV) == p-1:
	maxxx = 0
	for i in xrange(p):
		maxelem = max(FM[:TV[0]])
		for j in xrange(1,p-1):
			maxelem = max(maxelem, max(FM[TV[j-1]:TV[j]]))
		maxxx = max(maxxx, maxelem)
	maxx = 0
	for i in xrange(p):
		maxelem = max(FM[:V[0]])
		for j in xrange(1,p-1):
			maxelem = max(maxelem, max(FM[V[j-1]:V[j]]))
		maxx = max(maxx, maxelem)
	if maxxx < maxx:
		for s in TV:
			print s,
		print
	else:
		for s in V:
			print s,
		print
elif len(V) == p-1:
	for s in V:
		print s,
	print
elif len(TV) == p-1:
	for s in TV:
		print s,
	print
else:
	for s in H:
		print s,
	print

File: codechef/contests/SEPT16/CHMTDV.cpp
/*input
2 2
7 8
4 8
*/
#include <bits/stdc++.h>

using namespace std;

#define ll long long

ll n, p;

// inline void print(vector<ll> a){
// 	 for(ll x: a) cerr << x << " ";
// 		cerr << endl;
// }

inline ll fits(ll mid, ll *sum, vector<ll>& v){
	// //cerr << "Checking for " << mid << endl;
	ll cursum = 0;
	v.clear();
	for(ll j = 0; j < n; j += 1){
		if(sum[j] > mid) return 1;// increase sum i.e., lo=mid+1
		if(cursum + sum[j] > mid){
			v.push_back((ll)j);
			cursum = sum[j];
		}else{
			cursum += sum[j];
		}
	}
	if(v.size() > p-1) return 1;
	return 2;
}

inline ll fits2(ll mid, ll **FM, vector<ll>& V){
	ll cursum[p];
	V.clear();
	memset(cursum, 0, sizeof(cursum[0])*p);
	for (ll i = 0; i < n; i += 1){
		for(ll j = 0; j < p; j += 1){
			if(FM[j][i] > mid){
				return 1;
			}
			if (cursum[j]+FM[j][i] > mid){
				V.push_back(i);
				for(ll k = 0; k < p; k += 1){
					cursum[k] = FM[k][i];
				}
				break;
			}else{
				cursum[j] += FM[j][i];
			}
		}
	}
	if (V.size() > p-1) return 1;
	return 2;
}

int main(void){
	cin >> n >> p;
	ll a[n][n];
	vector<ll> H, V;
	for(ll i = 0; i < n; i += 1){
		for(ll j = 0; j < n; j += 1){
			cin >> a[i][j];
		}
	}
	ll *rsum = new ll[n];
	memset(rsum, 0, sizeof(rsum[0])*n);
	for(ll i = 0; i < n; i += 1){
		for(ll j = 0; j < n; j += 1){
			rsum[i] += a[i][j];
		}
	}
	//cerr << "Enetering to first Binary Search" << endl;
	ll lo = 0, hi = (ll)pow(10,15), mid = 0;
	while (lo <= hi){
		if(mid == (lo+hi)/2) break;
		mid = (hi+lo)/2;
		// //cerr << "mid: " << mid << endl;
		ll ans = fits(mid, rsum, H);
		if(ans == 2) hi = mid;
		else lo = mid+1;
	}
	if(lo > hi) swap(lo, hi);
	if(fits(lo, rsum, H) == 2){
		//cerr << "Rowsum: " << lo << endl;
	}else{
		fits(hi, rsum, H);
		//cerr << "Rowsum: " << hi << endl;
	}
	//cerr << "H: ";print(H);

	ll **FM = new ll*[p];
	for(ll i = 0; i < p; i += 1){
		FM[i] = new ll[n];
		fill_n(FM[i], n, 0);
	}

	// line below is wiping out much more memory than needed
	
	//cerr << "Memory FM set to 0" << endl;
	for(ll col = 0; col < n; col += 1){
		for(ll row = 0; row < p-1; row += 1){
			//cerr << "col: " << col << " row: " << row << endl;
			if(row == 0){
				for(ll crow = 0; crow < H[row]; crow += 1){
					FM[row][col] += a[crow][col];
				}
			}else{
				for(ll crow = H[row-1]; crow < H[row]; crow += 1){
					FM[row][col] += a[crow][col];
				}
				if(row != 0 and row == p-2){
					//cerr << "into the last" << endl;
					//cerr << "H: "; print(H);	
					for(ll crow = H[row]; crow < n; crow += 1){
						FM[row+1][col] += a[crow][col];
					}
					//cerr << "Sigh" << endl;
				}
			}
		}
	}
	//cerr << "Created FM " << endl;
	ll *csum = new ll[n];
	memset(csum, 0, sizeof(csum[0])*n);
	for(ll i = 0; i < p; i += 1){
		for(ll j = 0; j < n; j +=1){
			csum[j] += FM[i][j];
		}
	}
	// cerr << "The Horizontally divided matrix goes here" << endl;
	// for(int i = 0; i < p; i += 1){
	// 	for(int j = 0; j < n; j += 1){
	// 		cerr << FM[i][j] << " ";
	// 	}
	// 	cerr << endl;
	// }
	lo = 0, hi = (ll)pow(10,15), mid = 0;
	while (lo <= hi){
		if(mid == (lo+hi)/2) break;
		mid = (lo+hi)/2;
		ll ans = fits2(mid, FM, V);
			if(ans == 2) hi = mid;
		else lo = mid+1;
	}
	if(lo > hi) swap(lo, hi);
	if(fits2(lo, FM, V) == 2){
		// No need to do anything
	}else{
		fits2(hi, FM, V);
	}

	for(ll x: H) cout << x << " ";
		cout << endl;
	for(ll x: V) cout << x << " ";
		cout << endl;

}